src/app/clusters/chime-server/chime-server.cpp

```diff
@@ -0,0 +1,185 @@
+
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/****************************************************************************'
+ * @file
+ * @brief Implementation for the Chime Server Cluster
+ ***************************************************************************/
+
+#include "chime-server.h"
+
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/CommandHandlerInterfaceRegistry.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/InteractionModelEngine.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/util.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Chime;
+using namespace chip::app::Clusters::Chime::Attributes;
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+ChimeServer::ChimeServer(EndpointId endpointId, ChimeDelegate & delegate) :
+    AttributeAccessInterface(MakeOptional(endpointId), Chime::Id), CommandHandlerInterface(MakeOptional(endpointId), Chime::Id),
+    mDelegate(delegate)
+{
+    mDelegate.SetChimeServer(this);
+}
+
+ChimeServer::~ChimeServer()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+    CommandHandlerInterfaceRegistry::Instance().UnregisterCommandHandler(this);
+}
+
+CHIP_ERROR ChimeServer::Init()
+{
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INTERNAL);
+    ReturnErrorOnFailure(CommandHandlerInterfaceRegistry::Instance().RegisterCommandHandler(this));
+    return CHIP_NO_ERROR;
+}
+
+// AttributeAccessInterface
+CHIP_ERROR ChimeServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == Chime::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case InstalledChimeSounds::Id:
+        return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+            uint8_t index  = 0;
+            CHIP_ERROR err = CHIP_NO_ERROR;
+
+            // runs through all the Chime sounds available one by one
+            do
+            {
+                Chime::Structs::ChimeSoundStruct::Type chimeSound;
+                if ((err = this->mDelegate.GetChimeSoundByIndex(index, chimeSound)) == CHIP_NO_ERROR)
```

Comment:
So what's the ownership of the "Name" field of that struct?  This code assumes that it's pointing into some callee-allocated buffer that the callee can then free ... when?

It would be better to have a struct inheriting from `ChimeSoundStruct::Type` that has storage for the name so the callee can just copy into that.  See what the mode base bits do for an example, or some of the thermostat bits.

Reply:
Redone that section in line with how mode base does it - not as efficient but safer I think